<a href="https://colab.research.google.com/github/busraatasoy/veri_bilimi_ve_makine_ogrenmesi_100gun_kamp/blob/main/Day_02.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Gün 2: Veri Tipleri (Data Types)

Kodları yazarken not almak istersek **hashtag işareti (#)** kullanılmaktadır.

In [94]:
# Yorum satırı oluşturmak için hastag işareti kullanılır.
# Okunur kod yazmak için tanımlar, açıklamalar için kullanılmaktadır.
# Ayrıca, yazılan kodun başına hastag işareti konulursa kod çalışmaz.

Herhangi bir sayı, değer, karakterin veri tipini öğrenmek için` type()` tanımlı fonksiyonu kullanılır.

In [95]:
type(10)

int

In [96]:
type(10*3)

int

In [97]:
# Age ataması ve matematil işlemi

age = 50

age * 5

250

In [98]:
# İşlem sonucunu vermez, sonucun veri tipini verir.

type(age * 5)

int

## **int (Integer - Tam Sayı)**

Tam sayıları ifade eder. Ondalık (virgüllü) kısmı yoktur.

In [99]:
type(5)

int

In [100]:
type(0)

int

In [101]:
type(-1000)

int

In [102]:
# Değer atama

x = -100

In [103]:
type(x)

int

## **float (Floating Point - Ondalıklı Sayı)**

Ondalıklı, yani virgüllü (küsuratlı) sayıları ifade eder.

In [104]:
#Pi sayısı

type(3.14)

float

In [105]:
#Atama yapalım

pi =3.14

In [106]:
pi * 2

6.28

In [107]:
x  #değer atamıştık

-100

In [108]:
x + pi

-96.86

In [109]:
x / pi

-31.84713375796178

## **Değişkenin Değerini Değiştirmek – Atama**

**Dikkat:**

Daha önce x' e -100 olarak değer atamıştık. Peki, bunu değiştirebilir miyiz?

In [110]:
# x -100 yerine 10 olacak şekilde atamasını değiştirelim

x = 10

In [111]:
# x ile işlem yapalım

x + 3

13

In [112]:
# x atamasını tekrar değiştirelim

x = 20

In [113]:
# x ile aynı işlemi yapalım

x + 3

23

**Peki, ne oldu ?**

- Değişken x, önce -100, sonra 10, en son 20 değerini aldı.

- Her yeni atama, önceki değeri geçersiz kılar ve x artık son değeri olan 20’yi temsil eder.

- Sonuç olarak x + 3, artık 23 sonucunu verir.



✅ Sebebi:

Python'da değişkenler (variables), bir veriyi tutan **`etiketler (labels)`** gibidir.
Bu etiketin tuttuğu veriyi (yani değerini) değiştirmek çok kolaydır.

- x = 10 dediğimizde, x artık 10’u temsil eder.

- Önceki değer (örneğin -100), artık kullanılmaz.

Bu esneklik, Python’un **`dinamik tür`** yapısına **`(dynamic typing)`** dayanır.

📌 Unutma:

- Bir değişkenin değerini istediğin zaman değiştirebilirsin.

- Değeri değiştirmek için sadece = (atama operatörü) kullanman yeterli.

- Her atama işlemi, değişkenin önceki değerini yok sayar.

**❓ Soru:**

`int` ve `float` ikisi de sayı olmasına rağmen neden iki farklı veri tipi olarak tanımlanmıştır?



In [114]:
myFloat = 1.3442423

In [115]:
type(myFloat)

float

**✅Cevap:**

 Programlama dilleri; farklı özellikleri olan veri tiplerini aynı sınıflar, sınıflandırmalar içerisinde tutmaya çalışırlar.

Programlama dillerinde her veri tipinin hafızada kapladığı alan, işlem türü, ve hassasiyet seviyesi farklıdır.

Bu nedenle tamsayı (int) ile ondalıklı sayı (float) birbirinden ayrı veri tipleri olarak tanımlanmıştır.

**💾 Hafıza (Bellek) Açısından Fark:**

- Bilgisayarlar verileri ikili sistemde (0 ve 1) işler.
- Bu yüzden sayılar bellekte `bitler (binary digits)` olarak saklanır.

**int (tamsayı)** → daha az yer kaplar, çünkü sadece tam değerleri saklar.

Örn: 10 sayısı bellekte doğrudan 2-3 byte ile saklanabilir.

**float (ondalıklı sayı)** → daha fazla yer kaplar, çünkü virgülden sonraki kısmı da saklamak gerekir.

Örn: 1.3442423 gibi bir sayı, bellekte yaklaşık 8 byte yer kaplar ve daha karmaşık yapıda saklanır (IEEE 754 standardı kullanılır).

**🎯 Neden Ayrı Veri Tipleri?**

**Hassasiyet Farkı**

- float, çok küçük ya da çok büyük ondalıklı sayıları temsil edebilir.

- int ise sadece tam sayılar için optimize edilmiştir.

**Bellek Kullanımı***

- int daha az yer kaplar → büyük veri işlemlerinde avantaj sağlar.

- float daha fazla yer kaplar ama daha esnek hesaplamalara izin verir.

**İşlem Türü**

Bilgisayar, int ve float ile yapılan işlemleri farklı işlemcilerle yürütür (bazı sistemlerde).

Bu da performans farklarına yol açar.

**🔍 Sonuç:**

int ve float ikisi de sayı türü olsa da;

- Farklı hassasiyet

- Farklı bellek kullanımı

- Farklı matematiksel işlem yapısı

sebebiyle Python ve diğer dillerde ayrı veri tipleri olarak tanımlanmışlardır.



##**Adlandırmalar Nasıl Yapılır?**

Programlamada değişken, fonksiyon veya sınıf adları verirken belirli adlandırma kurallarına uymak hem okunabilirliği artırır hem de kodun profesyonel görünmesini sağlar.

**✅ İyi Adlandırma Kuralları:**

- Anlamlı ve okunabilir olmalı

- Rakamla başlamamalı

- Boşluk içeremez

- Türkçe karakterler (ç,ğ,ı,ö,ş,ü) kullanılmaz

- Özel karakterler (@, !, % vb.) kullanılmaz

- Küçük harf, büyük harf veya alt çizgi kullanılabilir

## **Adlandırma Biçimleri (Naming Conventions)**

**1. camelCase**

İlk kelime küçük harf, sonraki kelimelerin baş harfi büyük olur.

👉 Genellikle JavaScript, bazı Python değişkenlerinde kullanılır.

Örnek:

myFloat
studentName

**2. snake_case**

Kelimeler alt tire (_) ile ayrılır.

👉 Python’da en çok tercih edilen stildir.

Örnek:

my_float
student_name

Python programlarında genellikle snake_case tercih edilir. Ancak, her iki kullanımda uygundur.

**❌ Kötü Adlandırma Örnekleri:**

1. Rakamla başlayamaz
- 1name = "Ali"

2. Türkçe karakter kullanılmaz

- öğrenciAdı = "Ayşe"  

3. Boşluk olamaz
- my float = 5         

## **Int vs Float İşlem Farkları**

**İki tam sayının bölme işlemi**

In [116]:
# Bölme işlemi

40 / 4

10.0

In [117]:
# Neden bu işlem float?

type(40 / 4)

float

In [118]:
40 / 9

4.444444444444445

In [119]:
40 / 13

3.076923076923077

**📝Not:**

- Python programlama dili kapsamında 2 tam sayının bölümü kesirli çıkma ihtimali örneklerde olduğu gibi yüksek olduğu için veri tipi float olarak kararlaştırılmıştır.

- Bu Python'a özel bir durumdur ve sonuç tam sayı çıksa da geçerlidir. Örneğin `4 / 2 işlemi`nin sonucu `2 tamsayısı` olmasına rağmen işlem sonucu `2.0` olarak float değerde olur.

- Java programlama dilinde ise bu geçerli değildir. Böyle bir kural olmadığı için `tamsayı / tamsayı` sonucundan bağımsız` tam sayı değeri` döner. Tam bölünmese dahi kesirli olarak ifade edilmez. Fakat bu mantık da küsüratlı işlem gerektiren örneğin banka sekötrü gibi işlem gören yerlerde hatalı veri sonuçlarına yol açar. Burada da tip seçimi önem arz eder.

İşlemlerde `4 / 2` yerine `4.0 / 2.0 ` yazılmalıdır. Farkı nedir? Java'da bu anlamlı bir farktır çünkü tamsayı yerine float bölümü yapılması istenir. Python'da ise bu kural sebebiyle her iki işlemin sonucu da aynı döner.

In [120]:
type(4)

int

In [121]:
type(4.0)

float

**İki tam sayının çarpma işlemi**

In [122]:
4 * 2

8

**İki tam sayının toplama işlemi**

In [123]:
4 + 2

6

**İki tam sayının çıkarma işlemi**

In [124]:
4 - 2

2

Görüldüğü gibi; iki tam sayı işlemlerinde bu kural sadece bölme işleminde geçerlidir. çarpma, toplama ve çıkarma işlemlerinde sonuç tam sayı olacağı için bu kural geçerli değildir.

**Float - Int(Tam Sayı) Çıkarma İşlemi**

Bu işlemlerde kesirli sayıdan çıkarma yaptığımız için sonuç kesirli yani float veri tipinde olur.

In [125]:
4.0 - 2

2.0

In [126]:
4.1 - 2

2.0999999999999996

In [127]:
4.2 - 2

2.2

In [128]:
4.3 - 2

2.3

In [129]:
4.4 - 2

2.4000000000000004

In [130]:
4.5 - 2

2.5

In [131]:
type(4.5 - 2)

float

**Dipnot:**

Bazı durumlarda `4.4 - 2 = 2.4` yerine `2.4000000000000004` gibi sonuçlar Python'da gelmektedir. Çok önemsenmeyecek küçük hatalar olabilir. Bunlar float yani kesirli sayıların ele alınmasından kaynaklı oluşmaktadır. Hesaplamalarda veya kullanıcıya bilgi gösterirken sonucun bir kısmını alabiliriz.

**Float - Int(Tam Sayı) Çarpma İşlemi**

Çarpma işlemlerinde de çıkarma işleminde oldupu gibi float veri tipinde sonuç verir.

In [132]:
# Float - tam sayı değeri

4.2 * 3

12.600000000000001

In [133]:
# İki float değeri

4.2 * 3.2

13.440000000000001

In [134]:
# Float - tam sayı çok işlemli

4.2 * 3 - 10

2.6000000000000014

Python'da yazılan işlem sıralamasından bağımsız olarak matemetiksel işlem önceliğine göre hesaplama yapılır.

In [135]:
2 - 10 * 2

-18

In [136]:
2 - (10 * 2)

-18

Örnekte görüldüğü gibi, yazılan işlem sırasından bağımsız olarak matematiksel işlem önceliğindeki önce çarpma sonra çıkarma işlemi yapılır. Parantezli yazılmasa dahi bu durum geçerlidir.

**Üs Alma / Üslü Yazma İşlemi**

İki yıldız işareti `(**)` ile sayıların üs alma işlemi kolaylıkla yapılabilir.

In [137]:
2 * 2 * 2 * 2 * 2

32

In [138]:
2 ** 5

32

In [139]:
2 ** 9

512

**Kalanını Bulma(Remainder) İşlemi**

Bölme işleminden kalan bulma yüzde` (%)` işareti kullanılarak yapılır.

In [140]:
10 % 3

1

In [141]:
# Tam bölme işlemlerinde kalan `sıfır(0)`dır.

10 % 2

0

In [142]:
10 % 4

2

## **String Veri Tipi**

Programlama'ya ilk başlayanların yazdığı yazıyı yazmaya başlayarak ilerleyim.

String ifadeler çift tırnak `("")` olarak yazılmalıdır. Aksi halde **SyntaxError hatası** yani geçersiz söz dizimi uyarısı verir. Çift tırnak içine ifadeyi aldığımız da problem çözülür.

In [143]:
# Kodu çalıştırırsak hatayı alırız

# hello world

In [144]:
# Çift tırnak içinde yazınca hatayı düzeltiriz

"hello world"

'hello world'

**`print()` komutu:**

`print()` komutu yani fonksiyonu işlem yapmak istediğimiz bilgiyi ve kodu ekrana yazdırır.

**Dikkat:**

Jupyter notebook kullanırken her yazılan kodun çıktısı ekrara doğrudan yazdırıldığı için bu fonksiyonu kullanmadık.

Fakat; PyCharm, Visual Studio Code gibi farklı ortamlarda kod yazılınca çıktıları görmek için bu komutu yazmamız gereklidir.

In [145]:
print("hello world")

hello world


In [146]:
# String veri tipini atama

name = "atıl"

In [147]:
# Veri tipi kontrolü

type(name)

str

Tek `'....'` ve çift tırnak `"...."` içine string ifade tanımlanabilir. Aynı çıktı sonucunu verir.

In [148]:
'atil restaurant'

'atil restaurant'

In [149]:
"atil restaurant"

'atil restaurant'

Ancak, bazı `kesme işareti(apostrophe)` kullanılması gereken durumlar olduğunda tek tırnak kullanımında hata almamak için çift tırnak tercih edilir.

In [150]:
"atil's restaurant"

"atil's restaurant"

In [151]:
# Kod çalıştırıldığında tek tırnak kullanımında hata alınır

# 'atil's restaurant'

Görüldüğü gibi, `SyntaxError hatası` alındı. Burada tek tırnak içinde yazılan kısma kadar string ifade olarak tanındı fakat kalan kısım hata verdi. İfade açık kalmış halde görünmektedir.

Bu tip durumların önüne geçmek için çift tırnak ile string tanımı yapılabilir.

In [152]:
# Tanımlanan değişken

name

'atıl'

In [153]:
 # name değişken atamasının değiştirilmesi

name = "atil"

In [154]:
name

'atil'

In [155]:
type(name)

str

In [156]:
name = "büşra atasoy"

In [157]:
name

'büşra atasoy'

**Değişkenlerin Özelliklerine Erişme**

- name değişkeni -> nokta -> tab tuşu basıldığında (name.) bu değişken için yapılacak tüm işlemlere ulaşılır.
- name değişkeni string ifade olarak tanımlı olduğu için stringlere uygulanan capitalize gibi method ve fonksiyonlara pratik olarak ulaşılır.

In [158]:
# İlk harfini büyük harf ile yazmak için;

name.capitalize()

'Büşra atasoy'

**Fonksiyonlarda Parantez Kullanımı ve Anlamı**

- Yazılan fonksiyonlara parantezler eklenirse () gibi çalıştır anlamındadır.
- Parantezsiz yazıldığında ise, ne olduğunu yazan bir yazı ekrana yansır. Yani; Sadece FONKSİYONUN KENDİSİNE referans verilir.
Çalıştırılmaz, sadece “bu bir fonksiyondur” der gibi ekrana yansır.

- Bazıları () parantez eklenmeden de çalıştırılabilir bu kuralın geçerli olmadığı fonksiyon harici durumlar olabilir. Parantezsiz çalışıyorsa → muhtemelen fonksiyon değil, bir veri ya da nesnedir.


In [159]:
name.capitalize

<function str.capitalize()>

In [160]:
x = 10

In [161]:
type(x)

int

x. ile hızlıca bakılırsa stringlere kıyasla int ile yapılacak daha az işlem olduğu görülür.

- Farklı veri tiplerinin olması bu yönüyle de önemlidir. Bölme yapmak stringde anlam ifade etmediği gibi, ilk harfi büyük yapma gibi işlemler integerda anlamsızdır.

- Bu sebeple, her veri tipinin farklı olması ve uygulanacak işlemlerinde farklı olması anlamlıdır.

## **String Methodları**

Eğer değişkenlere uygulanacak işlemlerin nasıl uygulanacağını ve ne anlama geldiğini bilmiyorsak Python dökümanından yararlanabiliriz.

-> [Python Documentation Link:](https://www.python.org/doc/)

Dökümanı incelediğinizde anlatım seviyesi ileri gelebilir. Bu durumda Python help yönetimini kullanarak da işlemler hakkında bilgi sahibi olabiliriz. Kısa açıklamalı olarak hızlıca erişim sağlayabiliriz.

In [162]:
# Değişken ve uygulanmak istenen işlem yazılır.

help(name.count)

Help on built-in function count:

count(...) method of builtins.str instance
    S.count(sub[, start[, end]]) -> int

    Return the number of non-overlapping occurrences of substring sub in
    string S[start:end].  Optional arguments start and end are
    interpreted as in slice notation.



In [163]:
# Bilgi almak istediğimiz için parantezsiz yazılır.

help(name.capitalize)

Help on built-in function capitalize:

capitalize() method of builtins.str instance
    Return a capitalized version of the string.

    More specifically, make the first character have upper case and the rest lower
    case.



###**Count() methodu:**

Yazılan değişken içindeki harfin sayımını yapar.

Değişken isminde olmayan harf sayımı istendiğinde sıfır "0" sonucunu döndürür.

In [164]:
name

'büşra atasoy'

In [165]:
# Kaç tane "a" var?

name.count("a")

3

In [166]:
# Kaç tane "b" var?

name.count("b")

1

In [167]:
# Kaç tane "c" var?
name.count("c")


0

###**upper() methodu:**

String ifadenin tüm harflerini büyük harf haline getirir.

In [168]:
name.upper()

'BÜŞRA ATASOY'

**Peki; yapılan tüm değişiklikler kayıtlı mı ?**

capitalize() methodu ile ilk harfi büyük dönüştürdük, uyguladık, şimdi de upper() ile hepsini büyük harfe çevirdik.

Ancak; name yazıp değişken ismini çağırdığımızda en başta atanan değerde olduğunu görebiliriz.


In [169]:
name

'büşra atasoy'

**Peki; neden böyle ?**

Bazı fonksiyonlar çalışılan değişkenin içeriğini doğrudan değiştirebilir , bazıları ise, orijinal veriyi değiştirmez, sadece geçici olarak işlem sonucu üretir.

Bu durumda yeni bir değişken ataması oluşturularak değişikliğin kaydedilmesi gereklidir. Böylece yeni değişken çağrıldığında değişiklik uygulanmış olur.



**🧭 Sonuç:**

Bir metod çalıştırıldığında:

- Eğer sonucu yeni bir değişkene atamazsan, orijinal veri değişmez.

- Python, özellikle string gibi “değiştirilemez (immutable)” veri tiplerinde orijinal veriyi korumayı tercih eder.

In [170]:
# Değişkenleri tanımlayalım

x = 10
y = 5

In [171]:
# Değişken işlem sonucu

x + y

15

In [172]:
# z adlı başka değişkene atama

z = x + y

In [173]:
# değişkeni çağırma

z

15

In [174]:
name

'büşra atasoy'

In [175]:
#upper methodunu farklı değişkene atama

nameUpper = name.upper()

In [176]:
# yeni değişkeni çağırma

nameUpper

'BÜŞRA ATASOY'

Görüldüğü üzere, yapılan işlemler yeni değişkene atanıp çağrıldığında işlemlerin uygulandığı görünür ve kayıtlı hale gelir.

Burada name değişkenin içeriği yine değişmez sadece yapılan diğer işlemler yeni değişken ile kayıtlı hale gelir.

In [177]:
name

'büşra atasoy'

In [178]:
print(name.upper())

BÜŞRA ATASOY


**Not:**

Eğer uygulanılan işlem bilgisi sadece kullanıcıya anlık göstermek için ise; print içine işlem yazılarak gösterilir.

Fakat, işlem kayıtlı olması ve kod devamında başka işlemlerde kullanılacak ise, yeni değişken ataması ile kayıtlı hale getirilmelidir.

In [179]:
help(name.upper)

Help on built-in function upper:

upper() method of builtins.str instance
    Return a copy of the string converted to uppercase.



**Açıklamasında yazdığı gibi:**

`**Return a copy**` of the string converted to uppercase.

Dizenin büyük harfe dönüştürülmüş bir kopyasını döndür.

###**len() methodu:**

- Length yani uzunluk bilgisi verir.

In [180]:
# name: büşra atasoy olarak kaç harfli?
# Boşluk karakterini de sayar.

len(name)

12

In [181]:
help(len)

Help on built-in function len in module builtins:

len(obj, /)
    Return the number of items in a container.



##**Kaçış Karakterleri-Escape Characters**

**Farklı Satırda Yazdırma İşlemi:**

- **Ters slash ve n harfi (\n)** kullanılarak yapılır.
- Hangi satırı altta yazılmasını istiyorsak onun başına yazılır.

In [182]:
print("büşra atasoy")

büşra atasoy


In [183]:
print("büşra \natasoy")

büşra 
atasoy


**Bir tablık boşluk bırakma İşlemi:**

- Tek boşluk yerine bir tablık boşluk bırakılır.

- **Ters slash ve t harfi (\t)** kullanılarak yapılır.

In [184]:
print("büşra \tatasoy")

büşra 	atasoy
